File: src/com/gmail/marcosav2010/communicator/module/fth/FileReceiveInfo.java
package com.gmail.marcosav2010.communicator.module.fth;

public class FileReceiveInfo {

	private String filename;
	private int blocks;
	private long firstArrivalTime = -1, lastArrivalTime = -1;

	FileReceiveInfo(String filename, int blocks) {
		this.filename = filename;
		this.blocks = blocks;
	}

	public String getFileName() {
		return filename;
	}

	public int getBlocks() {
		return blocks;
	}

	public boolean isSingle() {
		return blocks == 1;
	}

	long getFirstArrivalTime() {
		return firstArrivalTime;
	}

	synchronized void setFirstArrivalTime() {
		if (firstArrivalTime == -1)
			lastArrivalTime = firstArrivalTime = System.currentTimeMillis();
	}

	synchronized boolean updateLastArrivalTime(long req) {
		long current = System.currentTimeMillis();
		long diff = current - lastArrivalTime;
		if (diff >= req) {
			lastArrivalTime = current;
			return true;

		} else
			return false;
	}
}

File: modules/file-transfer/src/main/java/com/gmail/marcosav2010/fth/FileReceiveInfo.kt
package com.gmail.marcosav2010.fth

class FileReceiveInfo internal constructor(val fileName: String, val blocks: Int = 0) {

    internal var firstArrivalTime: Long = -1
        private set

    private var lastArrivalTime: Long = -1

    val isSingle: Boolean get() = blocks == 1

    @Synchronized
    fun setFirstArrivalTime() {
        if (firstArrivalTime == -1L) {
            firstArrivalTime = System.currentTimeMillis()
            lastArrivalTime = firstArrivalTime
        }
    }

    @Synchronized
    fun updateLastArrivalTime(req: Long): Boolean {
        val current = System.currentTimeMillis()
        val diff = current - lastArrivalTime
        return if (diff >= req) {
            lastArrivalTime = current
            true
        } else false
    }
}